h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -546,6 +600,8 @@ public void reset() {
           _currentHyperparamIndices = null;
           _visitedPermutations.clear();
           _visitedPermutationHashes.clear();
+          _filterFunctions.stream().filter(fun -> fun instanceof KeepOnlyFirstMatchFilterFunction).forEach(fun -> ((KeepOnlyFirstMatchFilterFunction)fun).reset());
```

Comment:
If I'm correct, and `decrementCounter` is useless, then I would create a `Resetable` interface and use this instead here.

Reply:
@seb-h2o I would not say `decrementCounter` is useless. You just suggested to change `0/1` representation to `false/true` one.  `Resetable` might be a good idea but we will want to use this interface for `HyperSpaceIterator` as well.

Reply:
> you just suggested to change 0/1 representation to false/true one

Not only, I also suggest that applying the function should be enough to determine if it was applied or not, no need for an additional method.
Note that in your impl:
```
            filterFunctions.stream()
                    .filter(fun1 -> {
                      return fun1 instanceof KeepOnlyFirstMatchFilterFunction && ((KeepOnlyFirstMatchFilterFunction) fun1)._baseMatchFunction.apply(permutationAsMap);
                    })
                    .forEach(fun -> ((KeepOnlyFirstMatchFilterFunction) fun).decrementCounter());
          }
          return skipPermutation;
```
the `_baseMatchFunction` is applied a second time...
In my impl. suggestion, once the function is matched, any further application skips the "matchFunction" evaluation, and returns `false`. So no need to force the internal state using a method like `decrement`.